src/System.Security.SecureString/src/System/Security/SecureString.Windows.cs

```diff
@@ -0,0 +1,361 @@
+
+            ProtectMemory(decryptedBuffer);
+        }
+
+        private void AppendCharCore(char c)
+        {
+            SafeBSTRHandle decryptedBuffer = null;
+            try
+            {
+                decryptedBuffer = UnprotectMemory();
+                EnsureCapacity(ref decryptedBuffer, _decryptedLength + 1);
+                decryptedBuffer.Write<char>((uint)_decryptedLength * sizeof(char), c);
+                _decryptedLength++;
+            }
+            finally
+            {
+                ProtectMemory(decryptedBuffer);
+            }
+        }
+
+        // clears the current contents. Only available if writable
+        private void ClearCore()
+        {
+            _decryptedLength = 0;
+            _encryptedBuffer.ClearBuffer();
+        }
+
+        private void DisposeCore()
+        {
+            if (_encryptedBuffer != null && !_encryptedBuffer.IsInvalid)
+            {
+                _encryptedBuffer.Dispose();
+                _encryptedBuffer = null;
+            }
+        }
+
+        private unsafe void InsertAtCore(int index, char c)
+        {
+            byte* bufferPtr = null;
+            SafeBSTRHandle decryptedBuffer = null;
+            try
+            {
+                decryptedBuffer = UnprotectMemory();
+                EnsureCapacity(ref decryptedBuffer, _decryptedLength + 1);
+
+                decryptedBuffer.AcquirePointer(ref bufferPtr);
+                char* pBuffer = (char*)bufferPtr;
+
+                for (int i = _decryptedLength; i > index; i--)
+                {
+                    pBuffer[i] = pBuffer[i - 1];
+                }
+                pBuffer[index] = c;
+                ++_decryptedLength;
+            }
+            finally
+            {
+                if (bufferPtr != null)
+                {
+                    decryptedBuffer.ReleasePointer();
+                }
+                ProtectMemory(decryptedBuffer);
+            }
+        }
+
+        private unsafe void RemoveAtCore(int index)
+        {
+            byte* bufferPtr = null;
+            SafeBSTRHandle decryptedBuffer = null;
+            try
+            {
+                decryptedBuffer = UnprotectMemory();
+                decryptedBuffer.AcquirePointer(ref bufferPtr);
+                char* pBuffer = (char*)bufferPtr;
+
+                for (int i = index; i < _decryptedLength - 1; i++)
+                {
+                    pBuffer[i] = pBuffer[i + 1];
+                }
+                pBuffer[--_decryptedLength] = (char)0;
+            }
+            finally
+            {
+                if (bufferPtr != null)
+                {
+                    decryptedBuffer.ReleasePointer();
+                }
+                ProtectMemory(decryptedBuffer);
+            }
+        }
+
+        private void SetAtCore(int index, char c)
+        {
+            SafeBSTRHandle decryptedBuffer = null;
+            try
+            {
+                decryptedBuffer = UnprotectMemory();
```

Comment:
Shouldn't Unprotect be outside the try?  If Unprotect throws we don't want to double-protect...